pkg/apis/gcp/storage/v1alpha1/types.go

```diff
@@ -0,0 +1,936 @@
+
+// Bucket is the Schema for the instances API
+// +k8s:openapi-gen=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="STORAGE_CLASS",type="string",JSONPath=".spec.storageClass"
+// +kubebuilder:printcolumn:name="LOCATION",type="string",JSONPath=".spec.location"
+// +kubebuilder:printcolumn:name="CLASS",type="string",JSONPath=".spec.classRef.name"
+// +kubebuilder:printcolumn:name="RECLAIM_POLICY",type="string",JSONPath=".spec.reclaimPolicy"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+type Bucket struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BucketSpec   `json:"spec,omitempty"`
+	Status BucketStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BucketList contains a list of GCPBuckets
+type BucketList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Bucket `json:"items"`
+}
+
+// Resource Interface implementation
+var _ corev1alpha1.Resource = &Bucket{}
+
+// ConnectionSecretName returns a secret name from the reference
+func (b *Bucket) ConnectionSecretName() string {
+	return util.IfEmptyString(b.Spec.ConnectionSecretNameOverride, b.Name)
+}
+
+// ConnectionSecret returns a connection secret for this bucket instance
+func (b *Bucket) ConnectionSecret() *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:       b.Namespace,
+			Name:            b.ConnectionSecretName(),
+			OwnerReferences: []metav1.OwnerReference{b.OwnerReference()},
+		},
+		Data: map[string][]byte{
+			corev1alpha1.ResourceCredentialsSecretEndpointKey: []byte(b.GetUID()),
+		},
+	}
+}
+
+// ObjectReference to this resource instance
+func (b *Bucket) ObjectReference() *corev1.ObjectReference {
+	return util.ObjectReference(b.ObjectMeta, util.IfEmptyString(b.APIVersion, APIVersion), util.IfEmptyString(b.Kind, BucketKind))
+}
+
+// OwnerReference to use this instance as an owner
+func (b *Bucket) OwnerReference() metav1.OwnerReference {
+	return *util.ObjectToOwnerReference(b.ObjectReference())
+}
+
+// IsAvailable for usage/binding
+func (b *Bucket) IsAvailable() bool {
+	return b.Status.IsReady()
+}
+
+// IsBound determines if the resource is in a bound binding state
+func (b *Bucket) IsBound() bool {
+	return b.Status.Phase == corev1alpha1.BindingStateBound
+}
+
+// SetBound sets the binding state of this resource
+func (b *Bucket) SetBound(state bool) {
+	if state {
+		b.Status.Phase = corev1alpha1.BindingStateBound
+	} else {
+		b.Status.Phase = corev1alpha1.BindingStateUnbound
+	}
+}
+
+// NewBucketSpec constructs Spec for this resource from the properties map
+func NewBucketSpec(p map[string]string) *BucketSpec {
+	var encryption *BucketEncryption
+	if v, found := p["encryptionDefaultKmsKeyName"]; found {
+		encryption = &BucketEncryption{DefaultKMSKeyName: v}
+	}
```

Comment:
In these cases, yes, we want to test, and if `not found` - the resulting `BucketEncryption` (as well as others) will be `nil`.

Reply:
Is there a reason we don't want `Encryption` being set to a `*BucketEncryption` with zero values for all its fields? I ask because (iirc) I took that route with my controllers.